pkg/sqlite/image.go

```diff
@@ -0,0 +1,239 @@
+		errs = append(errs, err)
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// LoadBundleFunc walks the directory. When it sees a `.clusterserviceversion.yaml` file, it
+// attempts to load the surrounding files in the same directory as a bundle, and stores them in the
+// db for querying
+func (i *ImageLoader) LoadBundleFunc() error {
+	path := i.directory
+	manifests := filepath.Join(path, "manifests")
+	metadata := filepath.Join(path, "metadata")
+
+	// Get annotations file
+	log := logrus.WithFields(logrus.Fields{"dir": i.directory, "file": metadata, "load": "annotations"})
+	files, err := ioutil.ReadDir(metadata)
+	if err != nil {
+		return fmt.Errorf("unable to read directory %s: %s", metadata, err)
+	}
+
+	annotationsFile := &registry.AnnotationsFile{}
+	for _, f := range files {
+		fileReader, err := os.Open(filepath.Join(metadata, f.Name()))
+		if err != nil {
+			return fmt.Errorf("unable to read file %s: %s", f.Name(), err)
+		}
+		decoder := yaml.NewYAMLOrJSONDecoder(fileReader, 30)
+		err = decoder.Decode(&annotationsFile)
+		if err != nil || *annotationsFile == (registry.AnnotationsFile{}) {
+			continue
+		} else {
+			log.Info("found annotations file searching for csv")
+		}
+	}
+
+	if *annotationsFile == (registry.AnnotationsFile{}) {
+		return fmt.Errorf("Could not find annotations.yaml file")
+	}
+
+	err = i.loadManifests(manifests, annotationsFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *ImageLoader) loadManifests(manifests string, annotationsFile *registry.AnnotationsFile) error {
+	log := logrus.WithFields(logrus.Fields{"dir": i.directory, "file": manifests, "load": "bundle"})
+
+	csv, err := i.findCSV(manifests)
+	if err != nil {
+		return err
+	}
+
+	if csv.Object == nil {
+		return fmt.Errorf("csv is empty: %s", err)
+	}
+
+	log.Info("found csv, loading bundle")
+
+	// TODO: Check channels against what's in the database vs in the bundle csv
+
+	bundle, err := loadBundle(csv.GetName(), manifests)
+	if err != nil {
+		return fmt.Errorf("error loading objs in directory: %s", err)
+	}
+
+	if bundle == nil || bundle.Size() == 0 {
+		return fmt.Errorf("no bundle objects found")
+	}
+
+	// set the bundleimage on the bundle
+	bundle.BundleImage = i.image
+
+	if err := bundle.AllProvidedAPIsInBundle(); err != nil {
+		return fmt.Errorf("error checking provided apis in bundle %s: %s", bundle.Name, err)
+	}
+
+	bcsv, err := bundle.ClusterServiceVersion()
+	if err != nil {
+		return fmt.Errorf("error getting csv from bundle %s: %s", bundle.Name, err)
+	}
+
+	packageManifest, err := translateAnnotationsIntoPackage(annotationsFile, bcsv)
+	if err != nil {
+		return fmt.Errorf("Could not translate annotations file into packageManifest %s", err)
+	}
+
+	if err := i.loadOperatorBundle(packageManifest, *bundle); err != nil {
+		return fmt.Errorf("Error adding package %s", err)
+	}
+
+	// Finally let's delete all the old bundles
+	if err = i.store.ClearNonDefaultBundles(packageManifest.PackageName); err != nil {
```

Comment:
this should be behind a flag at least - it limits our options for transitioning (i.e. if I add an image bundle to an existing database, I lose all of my operator history...)

Reply:
Are you imagining that we want to support a workflow like that? imo if you are using `opm` it means you are just getting the new stuff that version of opm provides. otherwise it will start getting really hard about reasoning about migrations, for example.

if they want an old catalog, they are still able to use `initialize`

Reply:
Is anything in the csv that isn't in the db (as far as function is concerned)?